### DISCUSSION TITLE
Shrink parameters and options

### DISCUSSION THREAD

**Discussion by anitagraham** [libvips/libvips#2711] [Help wanted]
Hi just a quick question about shrink:

```
#shrink(hshrink, vshrink, **opts) ⇒ [Vips::Image]

    hshrink (Float) —   Horizontal shrink factor
    vshrink (Float) —   Vertical shrink factor
    opts (Hash) —    Set of options

Options Hash:
    :xshrink (Float) —   Horizontal shrink factor
    :yshrink (Float) —    Vertical shrink factor
```

Do we need both `hrshink` and `:xshrink`?
How are they different?

Recently I am seeing errors like:
```
GLib-GObject-WARNING **: 19:56:12.404: value "0.200000" of type 'gdouble' is invalid or out of range for property 'hshrink' of type 'gdouble'
GLib-GObject-WARNING **: 19:56:12.404: value "0.170000" of type 'gdouble' is invalid or out of range for property 'vshrink' of type 'gdouble'
GLib-GObject-WARNING **: 19:56:12.404: value "0.200000" of type 'gdouble' is invalid or out of range for property 'xshrink' of type 'gdouble'
GLib-GObject-WARNING **: 19:56:12.404: value "0.170000" of type 'gdouble' is invalid or out of range for property 'yshrink' of type 'gdouble'
WARN DRAGONFLY: caught error - shrink: parameter hshrink not set
shrink: parameter vshrink not set
```

Enlightenment is welcome.

Thanks, Anita

**Comment by jcupitt** [comment]
Hi @anitagraham,

libvips supports a `DEPRECATED` flag on operation arguments -- these are arguments that still work (to keep API compatibility) but which have been superseded and are no longer documented. 

Way back, `shrink` used to have arguments called `xshrink` and `yshrink`. The rest of the API uses `h` and `v` to mean the horizontal and vertical components of operations, so as part of the vips8 rewrite we deprecated `xshrink` and `yshrink` and added `hshrink` and `vshrink`. If you check the docs, they should be correct:

https://www.rubydoc.info/gems/ruby-vips/Vips/Image#shrink-instance_method

https://www.libvips.org/API/current/libvips-resample.html#vips-shrink

Where did you get the docs you quoted from? It probably needs fixing. 

> Recently I am seeing errors like:

`shrink` is a very fast and simple box filter -- all it does is set each output pixel to the average of the corresponding (`hshrink` x `vshrink`) box in the input image. The idea is that this very simple operation can be trivially vectorized and it makes a good first stage in a larger image downsizing pipeline (this is what the `resize` operation does). Obviously values less than 1 make no sense, hence the warnings. 

It sounds like dragonfly is maybe calling `shrink` when really it should call `resize`. In fact it probably shouldn't use `resize` either -- the main high-level image resizing entrypoint is `thumbnail`. I'd report a bug there.

The docs have some notes on this:

https://www.libvips.org/API/current/libvips-resample.html#libvips-resample.description

Though now I look they don't mention `thumbnail`. I'll add a para.

**Comment by jcupitt** [comment]
libvips docs update: 

https://github.com/libvips/libvips/commit/ede67090cb59c8cd53775424ed090e89147fa131

I had a quick look through dragonfly and it seems OK -- I think tomasc is using the API correctly.

Did you get the errors from calling `shrink` yourself? I would use `thumbnail`, if at all possible.

Tiny benchmark:

```
$ vipsheader nina.jpg 
nina.jpg: 6048x4032 uchar, 3 bands, srgb, jpegload
$ /usr/bin/time -f %M:%e vips thumbnail nina.jpg x.jpg 605
54080:0.07
$ /usr/bin/time -f %M:%e vips shrink nina.jpg x.jpg 10 10
93584:0.17
```

So `thumbnail` is 2x faster, half the memory use, and much better quality.

**Comment by anitagraham** [comment]
Yes, I  called shrink myself – I’ve just got back to the libvips “project” after a couple of months doing other work.

 

Thanks for the help.

 

From: John Cupitt ***@***.***>
Reply to: libvips/libvips ***@***.***>
Date: Sunday, 6 March 2022 at 7:55 pm
To: libvips/libvips ***@***.***>
Cc: Anita Graham ***@***.***>, Mention ***@***.***>
Subject: Re: [libvips/libvips] Shrink parameters and options (Discussion #2711)

 

libvips docs update:

ede6709

I had a quick look through dragonfly and it looks OK -- I think tomasc is using the API correctly.

Did you get the errors from calling shrink yourself? I would use thumbnail, if at all possible.

Tiny benchmark:
$ vipsheader nina.jpg 
nina.jpg: 6048x4032 uchar, 3 bands, srgb, jpegload
***@***.*** ~/pics $ /usr/bin/time -f %M:%e vips thumbnail nina.jpg x.jpg 605
54080:0.07
***@***.*** ~/pics $ /usr/bin/time -f %M:%e vips shrink nina.jpg x.jpg 10 10
93584:0.17
So thumbnail is 2x faster, half the memory use, and much better quality.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
